src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,73 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private List<Product> products;
+    private long maxId;
+    private Object lock = new Object();
+
+    public ArrayListProductDao(){
+        this.products = new ArrayList<>();
+        this.maxId = 0;
+        saveSampleProducts();
+    }
+
     @Override
-    public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+    public Product getProduct(Long id) throws NoSuchElementException {
+        return products.stream()
+                .filter(product -> id.equals(product.getId()))
+                .findAny()
+                .get();
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        return products.stream()
+                .filter(product -> product.getPrice() != null)
+                .filter(product -> product.getStock() > 0)
```

Comment:
Why not to make single filter with 2 conditions? It will increase performance a bit in large streams and also such approach is more popular (up to you to change or not).